Calibration/HcalCalibAlgos/plugins/HcalIsoTrkAnalyzer.cc

```diff
@@ -253,15 +270,24 @@ void HcalIsoTrkAnalyzer::analyze(const edm::Event& iEvent, const edm::EventSetup
   edm::Handle<reco::BeamSpot> beamSpotH;
   iEvent.getByToken(tok_bs_, beamSpotH);
   math::XYZPoint leadPV(0,0,0);
-  if (recVtxs.isValid() && recVtxs->size()>0 && !((*recVtxs)[0].isFake())) {
-    leadPV = math::XYZPoint( (*recVtxs)[0].x(),(*recVtxs)[0].y(), (*recVtxs)[0].z() );
-  } else if (beamSpotH.isValid()) {
+  t_goodPV = 0;
+  if (recVtxs.isValid() && recVtxs->size()>0) {
+    for (unsigned int k=0; k<recVtxs->size(); ++k) {
+      if (!((*recVtxs)[k].isFake()) && ((*recVtxs)[k].ndof() > 4)) {
+	if (t_goodPV == 0) leadPV = math::XYZPoint((*recVtxs)[k].x(),(*recVtxs)[k].y(),(*recVtxs)[k].z());
+	t_goodPV++;
+      }
+    }
+  } 
+  if (t_goodPV == 0 && beamSpotH.isValid()) {
     leadPV = beamSpotH->position();
   }
```

Comment:
@bsunanda IIUC the logic is the following: you want to
* define a primary vertex selection: being not fake and with at least 5 `dof`
* find the first occurrence of a primary vertex passing the selection in a valid non-zero-sized collection of reconstructed vertices, and extract its position
* enumerate all the vertices in the collection passing the selection;
* if no selection passed, use the BeamSpot position.

Maybe the code is not optimal here, but can go in, as of now.